File: crates/globstar/src/engine/thompson.rs
use crate::ast::CharClass;
use crate::engine::ops::{Op, OpProgram};

pub(crate) type StateId = u32;

const UNSET: StateId = StateId::MAX;

#[derive(Clone, Debug)]
pub(crate) enum Trans {
    /// Accept state. A thread here at end of input means a match.
    Match,
    /// Consume byte `b`.
    Byte { b: u8, next: StateId },
    /// Consume one byte in `class`.
    Class {
        class: Box<CharClass>,
        next: StateId,
    },
    /// Consume one non-separator byte.
    AnyNonSep { next: StateId },
    /// Consume any byte, separators included.
    AnyByte { next: StateId },
    /// Consume exactly one separator.
    Sep { next: StateId },
    /// ε fork. Both targets are taken during ε-closure.
    Split { a: StateId, b: StateId },
    /// ε jump to `next`.
    Jump { next: StateId },
    /// ε jump that dies when the thread is at a segment start and the next
    /// byte is `.`. Guards `*`'s zero-match exit under `dot=false`.
    DotGuard { next: StateId },
}

#[derive(Debug)]
pub(crate) struct Thompson {
    pub(crate) states: Vec<Trans>,
    pub(crate) initial: StateId,
    pub(crate) accept: StateId,
}

impl Thompson {
    pub(crate) fn compile(program: &OpProgram, dot: bool) -> Self {
        let mut builder = Builder::new(program.case_insensitive, dot);
        let (initial, tails) = builder.compile_ops(&program.ops);
        let accept = builder.alloc(Trans::Match);
        for st in tails {
            builder.patch(st, accept);
        }
        Self {
            states: builder.states,
            initial,
            accept,
        }
    }

    /// Per-state ε-closure bitmaps. `result[s * n_words .. (s+1) * n_words]`
    /// holds the leaf states reachable from `s` over Split/Jump edges.
    pub(crate) fn static_closures(&self, n_words: usize) -> Vec<u64> {
        let n = self.states.len();
        let mut closures = vec![0u64; n * n_words];
        let mut seen = vec![false; n];

        // High bit marks the exit phase (children done, fold their closures).
        const EXIT_BIT: u32 = 1 << 31;
        let mut stack: Vec<u32> = Vec::new();

        for root in 0..n {
            if seen[root] {
                continue;
            }
            stack.push(root as u32);

            while let Some(item) = stack.pop() {
                if item & EXIT_BIT != 0 {
                    let s = (item & !EXIT_BIT) as usize;
                    let s_base = s * n_words;
                    match &self.states[s] {
                        Trans::Split { a, b } => {
                            let a_base = (*a as usize) * n_words;
                            let b_base = (*b as usize) * n_words;
                            for j in 0..n_words {
                                closures[s_base + j] = closures[a_base + j] | closures[b_base + j];
                            }
                        }
                        Trans::Jump { next } => {
                            let n_base = (*next as usize) * n_words;
                            closures.copy_within(n_base..n_base + n_words, s_base);
                        }
                        _ => {
                            closures[s_base + (s >> 6)] = 1u64 << (s & 63);
                        }
                    }
                    continue;
                }

                let s = item as usize;
                if seen[s] {
                    continue;
                }
                seen[s] = true;
                stack.push((s as u32) | EXIT_BIT);

                match &self.states[s] {
                    Trans::Split { a, b } => {
                        if !seen[*a as usize] {
                            stack.push(*a);
                        }
                        if !seen[*b as usize] {
                            stack.push(*b);
                        }
                    }
                    Trans::Jump { next } if !seen[*next as usize] => {
                        stack.push(*next);
                    }
                    _ => {}
                }
            }
        }

        closures
    }

    /// Packed [`Trans::DotGuard`] expansions, one record of `1 + n_words`
    /// words per guard. A record holds the guard's state id, then the bitmap
    /// it releases when it passes. Chained guards are folded in, so the run
    /// loop expands every active guard in a single pass.
    pub(crate) fn guard_expansions(&self, closures: &[u64], n_words: usize) -> Box<[u64]> {
        let guards: Vec<usize> = self
            .states
            .iter()
            .enumerate()
            .filter(|(_, t)| matches!(t, Trans::DotGuard { .. }))
            .map(|(s, _)| s)
            .collect();
        if guards.is_empty() {
            return Box::new([]);
        }

        let stride = 1 + n_words;
        let mut recs = vec![0u64; guards.len() * stride];
        for (i, &g) in guards.iter().enumerate() {
            let Trans::DotGuard { next } = &self.states[g] else {
                unreachable!()
            };
            let rec = &mut recs[i * stride..(i + 1) * stride];
            rec[0] = g as u64;
            let base = *next as usize * n_words;
            rec[1..].copy_from_slice(&closures[base..base + n_words]);
        }

        let mut changed = true;
        while changed {
            changed = false;
            for i in (0..guards.len()).rev() {
                for (j, &gj) in guards.iter().enumerate() {
                    if i == j || recs[i * stride + 1 + (gj >> 6)] & (1u64 << (gj & 63)) == 0 {
                        continue;
                    }
                    for w in 0..n_words {
                        let merged = recs[i * stride + 1 + w] | recs[j * stride + 1 + w];
                        if merged != recs[i * stride + 1 + w] {
                            recs[i * stride + 1 + w] = merged;
                            changed = true;
                        }
                    }
                }
            }
        }
        recs.into_boxed_slice()
    }
}

struct Builder {
    states: Vec<Trans>,
    case_insensitive: bool,
    dot: bool,
}

impl Builder {
    fn new(case_insensitive: bool, dot: bool) -> Self {
        Self {
            states: Vec::with_capacity(32),
            case_insensitive,
            dot,
        }
    }

    fn alloc(&mut self, t: Trans) -> StateId {
        let id = self.next_id();
        self.states.push(t);
        id
    }

    fn next_id(&self) -> StateId {
        self.states.len() as StateId
    }

    fn patch(&mut self, state: StateId, target: StateId) {
        match &mut self.states[state as usize] {
            Trans::Match => panic!("cannot patch a Match state"),
            Trans::Byte { next, .. }
            | Trans::Class { next, .. }
            | Trans::AnyNonSep { next, .. }
            | Trans::AnyByte { next, .. }
            | Trans::Sep { next, .. }
            | Trans::Jump { next }
            | Trans::DotGuard { next } => {
                if *next == UNSET {
                    *next = target;
                }
            }
            // Every Split allocation site sets `a` eagerly; only `b` can
            // dangle (the zero-match exit of a dot=true star).
            Trans::Split { b, .. } => {
                if *b == UNSET {
                    *b = target;
                }
            }
        }
    }

    fn compile_ops(&mut self, ops: &[Op]) -> (StateId, Vec<StateId>) {
        if ops.is_empty() {
            let s = self.alloc(Trans::Jump { next: UNSET });
            return (s, vec![s]);
        }
        let (entry, mut pending_tails) = self.compile_op(&ops[0]);
        for op in &ops[1..] {
            let (op_entry, op_tails) = self.compile_op(op);
            for &tail in &pending_tails {
                self.patch(tail, op_entry);
            }
            pending_tails = op_tails;
        }
        (entry, pending_tails)
    }

    fn compile_op(&mut self, op: &Op) -> (StateId, Vec<StateId>) {
        match op {
            Op::Lit(bytes) => self.compile_lit(bytes),
            Op::AnyChar => self.compile_any_non_sep(),
            Op::Star => self.compile_star(),
            Op::Class(class) => self.compile_class(class),
            Op::Sep => self.compile_sep(),
            Op::SepRun => self.compile_sep_run(),
            Op::LeadingSeps => self.compile_leading_seps(),
            Op::OptSegmentsSlash => self.compile_oss(),
            Op::SlashAnything => self.compile_slash_anything(),
            Op::GlobstarAny => self.compile_globstar_any(),
            Op::Alternation(branches) => self.compile_alternation(branches),
            Op::Globstar => unreachable!("raw Globstar must be folded by lowering"),
        }
    }

    fn compile_lit(&mut self, bytes: &[u8]) -> (StateId, Vec<StateId>) {
        debug_assert!(
            !bytes.is_empty(),
            "empty Lit should not exist post-lowering"
        );
        let entry = self.alloc_lit_byte(bytes[0]);
        let mut prev = entry;
        for &b in &bytes[1..] {
            let s = self.alloc_lit_byte(b);
            self.patch(prev, s);
            prev = s;
        }
        (entry, vec![prev])
    }

    fn alloc_lit_byte(&mut self, b: u8) -> StateId {
        let class = self
            .case_insensitive
            .then(|| CharClass::ci_letter(b))
            .flatten();
        match class {
            Some(class) => self.alloc(Trans::Class {
                class: Box::new(class),
                next: UNSET,
            }),
            None => self.alloc(Trans::Byte { b, next: UNSET }),
        }
    }

    fn compile_any_non_sep(&mut self) -> (StateId, Vec<StateId>) {
        let s = self.alloc(Trans::AnyNonSep { next: UNSET });
        (s, vec![s])
    }

    fn compile_star(&mut self) -> (StateId, Vec<StateId>) {
        let entry = self.next_id();
        let body = entry + 1;
        if self.dot {
            self.alloc(Trans::Split { a: body, b: UNSET });
            self.alloc(Trans::AnyNonSep { next: entry });
            (entry, vec![entry])
        } else {
            let dot_guard = entry + 2;
            self.alloc(Trans::Split {
                a: body,
                b: dot_guard,
            });
            self.alloc(Trans::AnyNonSep { next: entry });
            self.alloc(Trans::DotGuard { next: UNSET });
            (entry, vec![dot_guard])
        }
    }

    fn compile_class(&mut self, class: &CharClass) -> (StateId, Vec<StateId>) {
        let s = self.alloc(Trans::Class {
            class: Box::new(class.clone()),
            next: UNSET,
        });
        (s, vec![s])
    }

    fn compile_sep(&mut self) -> (StateId, Vec<StateId>) {
        let entry = self.alloc(Trans::Sep { next: UNSET });
        (entry, vec![entry])
    }

    fn compile_sep_run(&mut self) -> (StateId, Vec<StateId>) {
        let tail_split = self.next_id();
        let entry = tail_split + 1;
        self.alloc(Trans::Split { a: entry, b: UNSET });
        self.alloc(Trans::Sep { next: tail_split });
        (entry, vec![tail_split])
    }

    fn compile_leading_seps(&mut self) -> (StateId, Vec<StateId>) {
        let entry = self.next_id();
        let loop_body = entry + 1;
        self.alloc(Trans::Split {
            a: loop_body,
            b: UNSET,
        });
        self.alloc(Trans::Sep { next: entry });
        (entry, vec![entry])
    }

    fn compile_oss(&mut self) -> (StateId, Vec<StateId>) {
        let entry = self.next_id();
        let (seg_body, seg_cont, seg_body_loop, sep_start, sep_tail) =
            (entry + 1, entry + 2, entry + 3, entry + 4, entry + 5);
        self.alloc(Trans::Split {
            a: seg_body,
            b: UNSET,
        });
        self.alloc(Trans::AnyNonSep { next: seg_cont });
        self.alloc(Trans::Split {
            a: seg_body_loop,
            b: sep_start,
        });
        self.alloc(Trans::AnyNonSep { next: seg_cont });
        self.alloc(Trans::Sep { next: sep_tail });
        self.alloc(Trans::Split {
            a: sep_start,
            b: entry,
        });
        (entry, vec![entry])
    }

    fn compile_slash_anything(&mut self) -> (StateId, Vec<StateId>) {
        let entry = self.next_id();
        let (post_sep, tail, tail_loop) = (entry + 1, entry + 2, entry + 3);
        self.alloc(Trans::Sep { next: post_sep });
        self.alloc(Trans::Split { a: entry, b: tail });
        self.alloc(Trans::Split {
            a: tail_loop,
            b: UNSET,
        });
        self.alloc(Trans::AnyByte { next: tail });
        (entry, vec![tail])
    }

    fn compile_globstar_any(&mut self) -> (StateId, Vec<StateId>) {
        let entry = self.next_id();
        let body = entry + 1;
        self.alloc(Trans::Split { a: body, b: UNSET });
        self.alloc(Trans::AnyByte { next: entry });
        (entry, vec![entry])
    }

    fn compile_alternation(&mut self, branches: &[Vec<Op>]) -> (StateId, Vec<StateId>) {
        debug_assert!(branches.len() >= 2);
        let mut branch_entries = Vec::with_capacity(branches.len());
        let mut branch_tails = Vec::new();
        for branch in branches {
            let (entry, tails) = self.compile_ops(branch);
            branch_entries.push(entry);
            branch_tails.extend(tails);
        }
        let mut entry = *branch_entries.last().unwrap();
        for &a in branch_entries[..branch_entries.len() - 1].iter().rev() {
            entry = self.alloc(Trans::Split { a, b: entry });
        }
        (entry, branch_tails)
    }
}

File: packages/globstar/src/engine/thompson.js
// Struct-of-arrays Thompson NFA builder for the PikeVM. JS port of the
// Rust module `crates/globstar/src/engine/thompson.rs`. Compiles to
// parallel `number[]` arrays
// instead of a `Trans` object per state — Builder skips ~N V8 object
// allocations per compile, freezes to typed arrays at the end, and
// hands PikeVm a ready-to-consume SoA shape.
//
// Per-state storage (during build, plain JS arrays):
//   tags[s]      number   T_BYTE | T_CLASS | …
//   nexts[s]     number   for byte-consumers/JUMP/DOT_GUARD: next state;
//                          for SPLIT: first branch (`a`)
//   byteVals[s]  number   T_BYTE byte value (else 0)
//   splitsB[s]  number   T_SPLIT second branch (else UNSET)
//   clsRefs[s]   Object   T_CLASS class struct (else null)
//
// Frozen output (consumed by PikeVm constructor):
//   info: Uint32Array(n) — packed tag/flags/byte/next per byte-consumer
//                          state (T_SPLIT/T_JUMP slots get tag=T_NULL,
//                          edges already absorbed by closures).
//   clsRefs: Array(n) | null
//   tags, nexts, splitsB: kept as number[] for `staticClosuresN` —
//                          dropped after closure computation.
//   initial, accept, n.

import {
  OP_LIT,
  OP_ANYCHAR,
  OP_STAR,
  OP_CLASS,
  OP_SEP,
  OP_SEP_RUN,
  OP_GLOBSTAR,
  OP_OPT_SEGMENTS_SLASH,
  OP_SLASH_ANYTHING,
  OP_GLOBSTAR_ANY,
  OP_LEADING_SEPS,
  OP_ALTERNATION,
} from "./ops/index.js";
import { ciLetter } from "../ast.js";

// NFA transition tags. `T_NULL` is a packed-runtime sentinel for ε-only
// states whose closures have already been absorbed by PikeVm.
export const T_MATCH = 0;
export const T_BYTE = 1;
export const T_CLASS = 2;
export const T_ANY_NON_SEP = 3;
export const T_ANY_BYTE = 4;
export const T_SEP = 5;
export const T_SPLIT = 6;
export const T_JUMP = 7;
export const T_DOT_GUARD = 8;
export const T_NULL = 9;

const UNSET = -1;

class Builder {
  constructor(caseInsensitive, dot) {
    this.caseInsensitive = caseInsensitive;
    // When false, `*` guards its zero-match exit with a T_DOT_GUARD.
    this.dot = dot;
    this.tags = [];
    this.nexts = [];
    this.byteVals = [];
    this.splitsB = [];
    this.clsRefs = []; // sparse — null entries common
  }

  _push(tag, next, byteVal, cls, splitB) {
    const id = this.tags.length;
    this.tags.push(tag);
    this.nexts.push(next);
    this.byteVals.push(byteVal);
    this.splitsB.push(splitB);
    this.clsRefs.push(cls);
    return id;
  }

  allocByte(b) {
    return this._push(T_BYTE, UNSET, b, null, UNSET);
  }
  allocClass(cls) {
    return this._push(T_CLASS, UNSET, 0, cls, UNSET);
  }
  allocAnyNonSep(next) {
    return this._push(T_ANY_NON_SEP, next, 0, null, UNSET);
  }
  allocAnyByte(next) {
    return this._push(T_ANY_BYTE, next, 0, null, UNSET);
  }
  allocSep(next) {
    return this._push(T_SEP, next, 0, null, UNSET);
  }
  allocSplit(a, splitB) {
    return this._push(T_SPLIT, a, 0, null, splitB);
  }
  allocJump(next) {
    return this._push(T_JUMP, next, 0, null, UNSET);
  }
  allocDotGuard(next) {
    return this._push(T_DOT_GUARD, next, 0, null, UNSET);
  }
  allocMatch() {
    return this._push(T_MATCH, 0, 0, null, UNSET);
  }

  allocLitByte(b) {
    if (this.caseInsensitive) {
      const cls = ciLetter(b);
      if (cls !== null) return this.allocClass(cls);
    }
    return this.allocByte(b);
  }

  patch(state, target) {
    const tag = this.tags[state];
    if (tag === T_MATCH) throw new Error("cannot patch a Match state");
    if (tag === T_SPLIT) {
      if (this.nexts[state] === UNSET) this.nexts[state] = target;
      else if (this.splitsB[state] === UNSET) this.splitsB[state] = target;
      return;
    }
    // BYTE / CLASS / ANY_NON_SEP / ANY_BYTE / SEP / JUMP / DOT_GUARD
    if (this.nexts[state] === UNSET) this.nexts[state] = target;
  }

  compileOps(ops) {
    if (ops.length === 0) {
      const s = this.allocJump(UNSET);
      return [s, [s]];
    }
    let entry = -1;
    let pendingTails = [];
    for (const op of ops) {
      const [opEntry, opTails] = this.compileOp(op);
      for (const t of pendingTails) this.patch(t, opEntry);
      pendingTails = opTails;
      if (entry === -1) entry = opEntry;
    }
    return [entry, pendingTails];
  }

  compileOp(op) {
    switch (op.kind) {
      case OP_LIT:
        return this.compileLit(op.bytes);
      case OP_ANYCHAR:
        return this.compileAnyNonSep();
      case OP_STAR:
        return this.compileStar();
      case OP_CLASS:
        return this.compileClass(op.cls);
      case OP_SEP:
        return this.compileSep();
      case OP_SEP_RUN:
        return this.compileSepRun();
      case OP_LEADING_SEPS:
        return this.compileLeadingSeps();
      case OP_OPT_SEGMENTS_SLASH:
        return this.compileOss();
      case OP_SLASH_ANYTHING:
        return this.compileSlashAnything();
      case OP_GLOBSTAR_ANY:
        return this.compileGlobstarAny();
      case OP_ALTERNATION:
        return this.compileAlternation(op.branches);
      case OP_GLOBSTAR: {
        // Lowering folds raw globstars away. Release safety net: an empty
        // class matches no byte, so a leak can never produce a false match.
        const s = this.allocClass({ neg: false, items: [] });
        return [s, [s]];
      }
    }
    throw new Error("compileOp: unreachable");
  }

  compileLit(bytes) {
    const entry = this.allocLitByte(bytes[0]);
    let prev = entry;
    for (let i = 1; i < bytes.length; i++) {
      const s = this.allocLitByte(bytes[i]);
      this.patch(prev, s);
      prev = s;
    }
    return [entry, [prev]];
  }

  compileAnyNonSep() {
    const s = this.allocAnyNonSep(UNSET);
    return [s, [s]];
  }

  compileStar() {
    const entry = this.allocSplit(UNSET, UNSET);
    const body = this.allocAnyNonSep(entry);
    this.nexts[entry] = body;
    if (!this.dot) {
      const dotGuard = this.allocDotGuard(UNSET);
      this.splitsB[entry] = dotGuard;
      return [entry, [dotGuard]];
    }
    // No dot protection: the Split's dangling exit is the fragment tail.
    return [entry, [entry]];
  }

  compileClass(cls) {
    const s = this.allocClass(cls);
    return [s, [s]];
  }

  compileSep() {
    const entry = this.allocSep(UNSET);
    return [entry, [entry]];
  }

  compileSepRun() {
    const tailSplit = this.allocSplit(UNSET, UNSET);
    const entry = this.allocSep(tailSplit);
    this.nexts[tailSplit] = entry;
    return [entry, [tailSplit]];
  }

  compileLeadingSeps() {
    const entry = this.allocSplit(UNSET, UNSET);
    const loopBody = this.allocSep(entry);
    this.nexts[entry] = loopBody;
    return [entry, [entry]];
  }

  compileOss() {
    const entry = this.allocSplit(UNSET, UNSET);
    const segBody = this.allocAnyNonSep(UNSET);
    const segCont = this.allocSplit(UNSET, UNSET);
    const segBodyLoop = this.allocAnyNonSep(segCont);
    const sepStart = this.allocSep(UNSET);
    const sepTail = this.allocSplit(UNSET, UNSET);

    this.nexts[segBody] = segCont;
    this.nexts[segCont] = segBodyLoop;
    this.splitsB[segCont] = sepStart;
    this.nexts[sepStart] = sepTail;
    this.nexts[sepTail] = sepStart;
    this.splitsB[sepTail] = entry;
    this.nexts[entry] = segBody;
    return [entry, [entry]];
  }

  compileSlashAnything() {
    const entry = this.allocSep(UNSET);
    const postSep = this.allocSplit(UNSET, UNSET);
    const tail = this.allocSplit(UNSET, UNSET);
    const tailLoop = this.allocAnyByte(tail);

    this.nexts[entry] = postSep;
    this.nexts[postSep] = entry;
    this.splitsB[postSep] = tail;
    this.nexts[tail] = tailLoop;
    return [entry, [tail]];
  }

  compileGlobstarAny() {
    const entry = this.allocSplit(UNSET, UNSET);
    const body = this.allocAnyByte(entry);
    this.nexts[entry] = body;
    return [entry, [entry]];
  }

  // Every constructible Alternation has >= 2 branches: the parser collapses
  // single-branch braces into literals (GLOB_SPEC §7.4) and the union
  // factorer only wraps >= 2 patterns.
  compileAlternation(branches) {
    const branchEntries = [];
    const branchTails = [];
    for (const branch of branches) {
      const [entry, tails] = this.compileOps(branch);
      branchEntries.push(entry);
      for (const t of tails) branchTails.push(t);
    }
    let nextState = -1;
    for (let i = branches.length - 2; i >= 0; i--) {
      const a = branchEntries[i];
      const b = nextState !== -1 ? nextState : branchEntries[i + 1];
      const s = this.allocSplit(a, b);
      nextState = s;
    }
    return [nextState, branchTails];
  }
}

// Build the SoA-form NFA. Returns the working data PikeVm needs to
// finish constructing its run state (closures, scratch, etc.).
export function compileThompson(program, dot) {
  const builder = new Builder(program.caseInsensitive, dot);
  const [initial, tails] = builder.compileOps(program.ops);
  const accept = builder.allocMatch();
  for (const st of tails) builder.patch(st, accept);

  const n = builder.tags.length;
  const tags = builder.tags;
  const nexts = builder.nexts;
  const splitsB = builder.splitsB;
  const byteVals = builder.byteVals;
  const clsRefs = builder.clsRefs;
  return {
    n,
    tags,
    nexts,
    splitsB,
    byteVals,
    clsRefs,
    initial,
    accept,
    dot,
  };
}
